### PR TITLE
exasol backend

### PR DESCRIPTION
**Pull request by Nicoretti**
- Add exasol db for testing
- Add exasol to pyproject.toml
- Add structure and package for exasol backend
- Add basic implementation of exasol backend
- Add basic documentation for exasol backend
- Add development notes

### CODE REVIEW

File: docs/backends/exasol.qmd [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Exasol
+
+[https://www.exasol.com](https://www.exasol.com)
+
+## Install
+
+Install Ibis and dependencies for the Druid backend:
```
Comment (tkilias) [inline docs/backends/exasol.qmd:7]:
```suggestion
Install Ibis and dependencies for the Exasol backend:
```

File: ibis/backends/exasol/__init__.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    from ibis.expr import datatypes as dt
+    from ibis.expr import operations as ops
+    from ibis.expr import types as ir
+
+
+class Backend(BaseAlchemyBackend, AlchemyCanCreateSchema):
+    name = "exasol"
+    compiler = ExasolCompiler
+    supports_create_or_replace = False
```
Comment (tkilias) [inline ibis/backends/exasol/__init__.py:22]: hm, for many things we support create or replace, we probably need to investigate what exactly this parameter means

File: ibis/backends/exasol/__init__.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    supports_python_udfs = False
```
Comment (tkilias) [inline ibis/backends/exasol/__init__.py:23]: uh, could be interesting later on

File: ibis/backends/exasol/__init__.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        user: str,
+        password: str,
+        host: str = "localhost",
+        port: int = 8563,
+        schema: str | None = None,
+        encryption: bool = True,
+        certificate_validation: bool = True,
+        encoding: str = "en_US.UTF-8",
+    ) -> None:
+        """TBD."""
```
Comment (tkilias) [inline ibis/backends/exasol/__init__.py:36]: don't forget this

File: ibis/backends/exasol/compiler.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    native_json_type = False
+    type_mapper = ExasolSQLType
+
+
+rewrites = ExasolExprTranslator.rewrites
+
+
+class ExasolCompiler(AlchemyCompiler):
+    translator_class = ExasolExprTranslator
+    support_values_syntax_in_select = False
```
Comment (tkilias) [inline ibis/backends/exasol/compiler.py:24]: this we probably need to check as well, might be supported

File: ibis/backends/exasol/tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+    from pathlib import Path
+
+T = TypeVar("T")
+
+
+@dataclass(frozen=True)
+class Option(Generic[T]):
```
Comment (tkilias) [inline ibis/backends/exasol/tests/conftest.py:22]: Looks familiar

File: ibis/backends/exasol/tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+            option.cli,
+            type=option.type,
+            help=option.help,
+        )
+
+
+class TestConf(ServiceBackendTest, RoundHalfToEven):
+    check_dtype = True
+    check_names = True
+    supports_arrays = True
```
Comment (tkilias) [inline ibis/backends/exasol/tests/conftest.py:186]: I think, we don't support arrays

File: ibis/backends/exasol/tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+    supports_arrays_outside_of_select = supports_arrays
+    supports_window_operations = True
+    supports_divide_by_zero = False
+    returned_timestamp_unit = "us"
+    supported_to_timestamp_units = {"s", "ms", "us"}
+    supports_floating_modulus = True
+    native_bool = True
+    supports_structs = True
+    supports_json = True
```
Comment (tkilias) [inline ibis/backends/exasol/tests/conftest.py:195]: this depends what it means, we have functions on varchar with json, but no native json type

File: ibis/backends/exasol/tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+        }
+        return ibis.exasol.connect(**kwargs)
+
+    @property
+    def test_files(self) -> Iterable[Path]:
+        return self.data_dir.joinpath("csv").glob("*.csv")
+
+    def _exaplus(self) -> str:
+        try:
+            binary = "/usr/opt/EXASuite-7/EXASolution-7.1.22/bin/Console/exaplus"
```
Comment (tkilias) [inline ibis/backends/exasol/tests/conftest.py:224]: why do you try first a hardcoded path?

File: ibis/backends/exasol/tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+            self._exaplus(),
+            "-c",
+            f"{EXASOL.host}:{EXASOL.port}",
+            "-u",
+            EXASOL.username,
+            "-p",
+            EXASOL.password,
+            "-f",
+            ddl_file,
+            # TODO: custom generated certificate,
```
Comment (tkilias) [inline ibis/backends/exasol/tests/conftest.py:266]: that is quite a lot of effort to add a custom generated certificate

**Review by tkilias — CHANGES_REQUESTED** [review]

File: ibis/backends/exasol/tests/conftest.py (hunk 9 above)
Comment (Nicoretti) [reply ibis/backends/exasol/tests/conftest.py:222]: because it have been significantly faster

File: ibis/backends/exasol/tests/conftest.py (hunk 10 above)
Comment (Nicoretti) [reply ibis/backends/exasol/tests/conftest.py:264]: Don't recall when i have put it there, for testing it can be skipped for now. Good point.

**Pull request closed without merging**